core/src/core/compact_block.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2018 The Grin Developers
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Compact Blocks.
+
+use consensus::VerifySortOrder;
+use core::block::{BlockHeader, Error};
+use core::{Output, ShortId, TxKernel};
+use ser::{self, read_and_verify_sorted, Readable, Reader, Writeable, Writer};
+
+/// Container for full (full) outputs and kernels and kern_ids for a compact block.
+#[derive(Debug, Clone)]
+pub struct CompactBlockBody {
+	/// List of full outputs - specifically the coinbase output(s)
+	pub out_full: Vec<Output>,
+	/// List of full kernels - specifically the coinbase kernel(s)
+	pub kern_full: Vec<TxKernel>,
+	/// List of transaction kernels, excluding those in the full list
+	/// (short_ids)
+	pub kern_ids: Vec<ShortId>,
+}
+
+impl CompactBlockBody {
+	fn validate(&self) -> Result<(), Error> {
+		self.verify_sorted()?;
+		Ok(())
+	}
+
+	// Verify everything is sorted in lexicographical order.
+	fn verify_sorted(&self) -> Result<(), Error> {
+		self.out_full.verify_sort_order()?;
+		self.kern_full.verify_sort_order()?;
+		self.kern_ids.verify_sort_order()?;
```

Comment:
We may replace last 2 lines with `self.kern_ids.verify_sort_order()`

Reply:
I actually prefer it with an explicit `Ok(())` at the end, to keep the various lines above all consistent.
There's also something in rust (I think `From` is only called automatically from within `try!`?) where errors will not necessarily be converted without the explicit `...?` syntax.
We don't need this everywhere but I think its better to just be consistent with this.